Fix the bug in THCTensor_(baddbmm) and ATen's addmm_cuda for strided views input

IvanYashchuk opened the pull request:
Fixes #42418. 

The problem was that the non-contiguous batched matrices were passed to `gemmStridedBatched`.

The following code fails on master and works with the proposed patch:
```python
import torch
x = torch.tensor([[1., 2, 3], [4., 5, 6]], device='cuda:0')
c = torch.as_strided(x, size=[2, 2, 2], stride=[3, 1, 1])
torch.einsum('...ab,...bc->...ac', c, c)
```

IvanYashchuk committed:
Fixed issue #42418.

The problem was that the wrong codepath was taken with unusual strides and not contiguous input in batched matrix multiplication.

ngimel commented:
Thank you, the fix looks good! Please add the test to the test suite. There's test_baddbmm in test_torch.py, but for some reason it's enabled only on the CPU. Can you try enabling it on cuda and adding a testcase for the behavior your enabled there?

IvanYashchuk commented:
Sure, I will do that.

IvanYashchuk committed:
Fixed a bug in prepare_matrix_for_cublas function used for torch.mm

IvanYashchuk committed:
Added tests for torch.mm and torch.bmm with strided input, results are compared to NumPy

IvanYashchuk commented:
While writing tests I found that `torch.mm` also doesn't work with this kind of strided input
```python
import torch
x = torch.tensor([[1., 2, 3], [4., 5, 6]], device='cuda:0')
c = torch.as_strided(x, size=[2, 2, 2], stride=[3, 1, 1])
torch.mm(c[0], c[0]) # Fails
```
The same problem was in `ATen/native/cuda/LinearAlgebra.cu:prepare_matrix_for_cublas` as in `THCTensor_(baddbmm)` that the input was not transformed into a contiguous array. I have fixed that.
I've added the tests both for `pytorch.mm` and `pytorch.bmm`. Correctness is checked via comparing to the NumPy result.

ngimel commented on test/test_torch.py:
```diff
@@ -17035,6 +17035,32 @@ def genf_float(x, y):
+        x = torch.tensor([[1., 2., 3.], [4., 5., 6.]], dtype=dtype, device=device)
+        new_shape = [2, 2, 2]
+        new_stride = [3, 1, 1]
+        sx = torch.as_strided(x, size=new_shape, stride=new_stride)
```
can you please use `compare_with_numpy` helper here to compare results? You don't need numpy arrays to have the same striding, so `.detach().cpu().numpy()` that `compare_with_numpy` does should be fine. To deal with multiple arguments, you can do (for torch_fn, and similarly for numpy_fn) `lambda x: torch.bmm(x,x)`. 
You don't need to compare results of torch.bmm with torch.mm, if you are comparing with numpy anyway

IvanYashchuk replied:
I've modified the test to use `compare_with_numpy` now.

ngimel reviewed (commented):
Thank you, this looks great! I have a small suggestion about the test.

IvanYashchuk committed:
Use compare_with_numpy in tests

ngimel reviewed (approved):
Awesome, thanks!

ngimel commented:
flake8 error is real

IvanYashchuk committed:
noqa: E731 on lambdas in the test

IvanYashchuk commented:
> flake8 error is real

Is it okay to ignore E731 because of assigned lambdas instead of def in the test?

ngimel commented:
Yeah, that's fine, you could also use lambdas directly as arguments, but it does not matter. Sorry to ask, but can you please rebase? We had an issue with docker images today, so CI on this PR is failing because it is against a bad base, and can't find docker images.

IvanYashchuk commented:
Alright, I did something wrong here. I've rebased onto master and pushed to branch and PR was automatically closed.

ngimel commented:
It somehow became 0-commit 0-line pull request, that's probably why it was closed. Something wrong with the rebase?

IvanYashchuk commented:
I've recovered the branch. Can this PR be reopened? I obviously don't have rights for this. I am sorry for the inconvenience.

ngimel commented:
Thank you!
